File: VB/AuthenticationOwin.Web/Startup.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Configuration
Imports System.Threading.Tasks
Imports Microsoft.Owin
Imports Microsoft.Owin.Security
Imports Microsoft.Owin.Security.Cookies
Imports Microsoft.Owin.Security.Facebook
Imports Microsoft.Owin.Security.Google
Imports Owin
Imports System.Net.Http
Imports System.Threading
Imports Microsoft.Owin.Security.MicrosoftAccount
Imports DevExpress.ExpressApp.Web
Imports System.Security.Claims
Imports Microsoft.AspNet.Identity

<Assembly: OwinStartup(GetType(AuthenticationOwin.Web.Startup))>

Namespace AuthenticationOwin.Web
    Public Class Startup
        Private Shared googleClientID As String = ConfigurationManager.AppSettings("GoogleClientID")
        Private Shared googleClientSecret As String = ConfigurationManager.AppSettings("GoogleClientSecret")
        Private Shared facebookClientID As String = ConfigurationManager.AppSettings("FacebookClientID")
        Private Shared facebookClientSecret As String = ConfigurationManager.AppSettings("FacebookClientSecret")
        Private Shared microsoftClientID As String = ConfigurationManager.AppSettings("MicrosoftClientID")
        Private Shared microsoftClientSecret As String = ConfigurationManager.AppSettings("MicrosoftClientSecret")

        Public Sub Configuration(ByVal app As IAppBuilder)
            app.UseCookieAuthentication(New CookieAuthenticationOptions With {
                                        .AuthenticationType = DefaultAuthenticationTypes.ApplicationCookie})

            app.UseExternalSignInCookie(DefaultAuthenticationTypes.ExternalCookie)

            If (Not String.IsNullOrEmpty(googleClientID)) AndAlso (Not String.IsNullOrEmpty(googleClientID)) Then
                app.UseGoogleAuthentication(New GoogleOAuth2AuthenticationOptions() With {.ClientId = googleClientID, .ClientSecret = googleClientSecret})
            End If
            If (Not String.IsNullOrEmpty(facebookClientID)) AndAlso (Not String.IsNullOrEmpty(facebookClientSecret)) Then
                Dim facebookAuthOptions As FacebookAuthenticationOptions = New FacebookAuthenticationOptions With {.AppId = facebookClientID, .AppSecret = facebookClientSecret, .UserInformationEndpoint = "https://graph.facebook.com/v2.4/me?fields=id,name,email"}
                app.UseFacebookAuthentication(facebookAuthOptions)
            End If
            If ((Not String.IsNullOrEmpty(microsoftClientID)) AndAlso (Not String.IsNullOrEmpty(microsoftClientSecret))) Then
                Dim microsoftAccountAuthenticationOptions As MicrosoftAccountAuthenticationOptions = New MicrosoftAccountAuthenticationOptions With {
                    .ClientId = microsoftClientID,
                    .ClientSecret = microsoftClientSecret,
                    .Provider = New MicrosoftAccountAuthenticationProvider() With {.OnAuthenticated = Function(context)
                                                                                                          Dim email = context.User("userPrincipalName")
                                                                                                          If (email IsNot Nothing) Then
                                                                                                              context.Identity.AddClaim(New Claim(ClaimTypes.Email, email.ToString()))
                                                                                                          End If
                                                                                                          Return Task.FromResult(0)
                                                                                                      End Function}}
                app.UseMicrosoftAccountAuthentication(microsoftAccountAuthenticationOptions)
            End If
        End Sub
    End Class
End Namespace

File: CS/AuthenticationOwin.Web/Startup.cs
using System;
using System.Configuration;
using System.Threading.Tasks;
using Microsoft.Owin;
using Microsoft.Owin.Security;
using Microsoft.Owin.Security.Cookies;
using Microsoft.Owin.Security.Facebook;
using Microsoft.Owin.Security.Google;
using Owin;
using System.Net.Http;
using System.Threading;
using Microsoft.Owin.Security.MicrosoftAccount;
using DevExpress.ExpressApp.Web;
using System.Security.Claims;
using Microsoft.AspNet.Identity;

[assembly: OwinStartup(typeof(AuthenticationOwin.Web.Startup))]

namespace AuthenticationOwin.Web {
    public class Startup {
        private static string googleClientID = ConfigurationManager.AppSettings["GoogleClientID"];
        private static string googleClientSecret = ConfigurationManager.AppSettings["GoogleClientSecret"];
        private static string facebookClientID = ConfigurationManager.AppSettings["FacebookClientID"];
        private static string facebookClientSecret = ConfigurationManager.AppSettings["FacebookClientSecret"];
        private static string microsoftClientID = ConfigurationManager.AppSettings["MicrosoftClientID"];
        private static string microsoftClientSecret = ConfigurationManager.AppSettings["MicrosoftClientSecret"];

        public void Configuration(IAppBuilder app) {
            app.UseCookieAuthentication(new CookieAuthenticationOptions {
                AuthenticationType = DefaultAuthenticationTypes.ApplicationCookie
            });

            app.UseExternalSignInCookie(DefaultAuthenticationTypes.ExternalCookie);

            if (!string.IsNullOrEmpty(googleClientID) && !string.IsNullOrEmpty(googleClientID)) {
                app.UseGoogleAuthentication(new GoogleOAuth2AuthenticationOptions() {
                    ClientId = googleClientID,
                    ClientSecret = googleClientSecret
                });
            }
            if(!string.IsNullOrEmpty(facebookClientID) && !string.IsNullOrEmpty(facebookClientSecret)) {
                FacebookAuthenticationOptions facebookAuthOptions = new FacebookAuthenticationOptions {
                    AppId = facebookClientID,
                    AppSecret = facebookClientSecret,
                    UserInformationEndpoint = "https://graph.facebook.com/v2.4/me?fields=id,name,email",
                    Scope = { "email" }
                };
                app.UseFacebookAuthentication(facebookAuthOptions);
            }
            if((!string.IsNullOrEmpty(microsoftClientID) && !string.IsNullOrEmpty(microsoftClientSecret))) {
                MicrosoftAccountAuthenticationOptions microsoftAccountAuthenticationOptions = new MicrosoftAccountAuthenticationOptions {
                    ClientId = microsoftClientID,
                    ClientSecret = microsoftClientSecret,
                    Provider = new MicrosoftAccountAuthenticationProvider() {
                        OnAuthenticated = (context) => {
                            var email = context.User["userPrincipalName"];
                            if(email != null) {
                                context.Identity.AddClaim(new Claim(ClaimTypes.Email, email.ToString()));
                            }
                            return Task.FromResult(0);
                        }
                    }
                };
                app.UseMicrosoftAccountAuthentication(microsoftAccountAuthenticationOptions);
            }
        }
    }
}
